commit a4c3be2e4209f2516233ce8c50ddaeb047a58501
Author:     Olivier Dony <odo@openerp.com>
AuthorDate: 2017-02-15 22:32:21 +0100
Commit:     Olivier Dony <odo@openerp.com>
CommitDate: 2017-02-15 22:32:37 +0100

    [FIX] sql_db: port fix from psycopg/psycopg2#459
    
    NUL characters must not be used in query parameters,
    as they will be ignored by libpq, being end-of-string
    characters.
    
    Preventing NULs avoids unexpected results from
    queries. It is only necessary with psycopg2
    versions before 2.7, which includes the upstream
    fix.

diff --git odoo/sql_db.py odoo/sql_db.py
index f2b2002..326ea44 100644
--- odoo/sql_db.py
+++ odoo/sql_db.py
@@ -47,6 +47,19 @@ psycopg2.extensions.register_type(psycopg2.extensions.new_type((700, 701, 1700,)
 
 
 import tools
+
+from tools import parse_version as pv
+if pv(psycopg2.__version__) < pv('2.7'):
+    from psycopg2._psycopg import QuotedString
+    def adapt_string(adapted):
+        """Python implementation of psycopg/psycopg2#459 from v2.7"""
+        if '\x00' in adapted:
+            raise ValueError("A string literal cannot contain NUL (0x00) characters.")
+        return QuotedString(adapted)
+
+    psycopg2.extensions.register_adapter(str, adapt_string)
+    psycopg2.extensions.register_adapter(unicode, adapt_string)
+
 from tools.func import frame_codeinfo
 from datetime import timedelta
 import threading
